pilopt/src/lib.rs

```diff
@@ -0,0 +1,416 @@
+            // TODO use a proper evaluator at some point,
+            // combine with constant_evalutaor
+            let mut values = expressions
+                .iter()
+                .filter(|e| e.repetitions > 0)
+                .flat_map(|e| e.values.iter())
+                .map(|e| match e {
+                    Expression::Number(n) => Some(n),
+                    _ => None,
+                });
+            let Some(Some(first)) = values.next() else { return None; };
+            if values.all(|x| x == Some(first)) {
+                Some(*first)
+            } else {
+                None
+            }
+        }
+        FunctionValueDefinition::Query(_) => None,
+    }
+}
+
+/// Simplifies multiplecations by zero and one.
+fn simplify_expressions<T: FieldElement>(pil_file: &mut Analyzed<T>) {
+    postvisit_expressions_in_pil_file_mut(pil_file, &mut |e| -> ControlFlow<()> {
+        simplify_expression_single(e);
+        ControlFlow::Continue(())
+    });
+}
+
+fn simplify_expression<T: FieldElement>(mut e: Expression<T>) -> Expression<T> {
+    postvisit_expression_mut(&mut e, &mut |e| -> ControlFlow<()> {
+        simplify_expression_single(e);
+        ControlFlow::Continue(())
+    });
+    e
+}
+
+fn simplify_expression_single<T: FieldElement>(e: &mut Expression<T>) {
+    if let Expression::BinaryOperation(left, op, right) = e {
+        if let (Expression::Number(l), Expression::Number(r)) = (left.as_ref(), right.as_ref()) {
+            if let Some(v) = match op {
+                BinaryOperator::Add => Some(*l + *r),
+                BinaryOperator::Sub => Some(*l - *r),
+                BinaryOperator::Mul => Some(*l * *r),
+                // TODO we might do some more operations later.
+                _ => None,
+            } {
+                *e = Expression::Number(v);
+                return;
+            }
+        }
+    }
+    if let Expression::UnaryOperation(op, inner) = e {
+        if let Expression::Number(inner) = **inner {
+            *e = Expression::Number(match op {
+                UnaryOperator::Plus => inner,
+                UnaryOperator::Minus => -inner,
+            });
+            return;
+        }
+    }
+    match e {
+        Expression::BinaryOperation(left, BinaryOperator::Mul, right) => {
+            if let Expression::Number(n) = left.as_mut() {
+                if *n == 0.into() {
+                    *e = Expression::Number(0.into());
+                    return;
+                }
+            }
+            if let Expression::Number(n) = right.as_mut() {
+                if *n == 0.into() {
+                    *e = Expression::Number(0.into());
+                    return;
+                }
+            }
+            if let Expression::Number(n) = left.as_mut() {
+                if *n == 1.into() {
+                    let mut tmp = Expression::Number(1.into());
+                    std::mem::swap(&mut tmp, right);
+                    std::mem::swap(e, &mut tmp);
```

Comment:
Multiple problems. The most obvious thing would be to use `*e = right`, but then we cannot move from right. Next would be `e = std::mem::take(right)`, but that doesn't work since `Expression` does not implement `Default`. `swap(e, right)` does not work because right is included in e. This is the best I could come up with.